### PR TITLE
Add player and strategy class

### PR DESCRIPTION
**Pull request by sabrasisler**

### CODE REVIEW

File: scrabbler/player.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            self.rack.remove('?')
+        else:
+            self.rack.remove(move.word[index])
+
+    def get_score(self):
+        """ getter for the player's score"""
+        return self.score
+
+
+class Strategy:
```
Comment (jfoley-yw) [inline scrabbler/player.py]: Can you move this class to its own file? I'm thinking it'd make more sense to raise a "not defined" error in this class' `choose_move` method (like in the Pacman assignments), and then create a class called something like `BaselineStrategy` that implements `Strategy` and defines the `choose_move` method the way you defined it below (all the strategy classes can go in one file).

File: scrabbler/player.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+class Player:
+    '''
+    Player Class represents a player in scrabble that takes in a midgame strategy and an optional endgame strategy. 
+    '''
+    def __init__ (self, midgame_strategy, endgame_strategy=None, name=None):
+        self.mid_strat = midgame_strategy
+        # If no end game strategy is specified, the midgame strategy will be applied as the endgame.
+        if endgame_strategy is None:
```
Comment (jfoley-yw) [inline scrabbler/player.py:8]: Nice 👍.

File: scrabbler/simulation.py [hunk 3]
```diff
@@ -18,26 +21,25 @@ class Simulation:
...
         self.board = self.game.board
-        # Rack starts out empty.
-        self.racks = [[], []]
-        self.scores = [0, 0]
-        # List of letters we can still pick from.
+        # List of letters we can still  pick from.
         self.bag = list(Simulation.LETTERS)
-        self.player = 0
+        # randomly choose which player goes first so that it varies during each simulation
+        self.player = random.randint(0,1)
```
Comment (jfoley-yw) [inline scrabbler/simulation.py:33]: 👍

File: scrabbler/simulation.py [hunk 4]
```diff
@@ -18,26 +21,25 @@ class Simulation:
...
         self.endgame = False
         self.times = []
+        self.players = (player1, player2)
 
     def simulate(self):
         # Keep playing until we're out of tiles or solutions.
         while self.exectute_turn():
             # switch whose turn it is
-            self.player = 1 - self.player
+            self.player= 1 - self.player
```
Comment (jfoley-yw) [inline scrabbler/simulation.py]: nit - can you add this space back?

File: scrabbler/simulation.py [hunk 5]
```diff
@@ -105,25 +111,19 @@ def make_move(self, move):
...
-            self.racks[self.player].remove(move.word[index])
+        self.players[self.player].update_score(move.score)
 
     def print_end_game_message(self):
-        print('\nGAME OVER!')
-        print("PLAYER 1 SCORE: %d ...... PLAYER 2 SCORE: %d" % (self.scores[0], self.scores[1]))
-        print('Average move-generation time:', sum(self.times) / len(self.times))
+        print("\nGAME OVER!")
+        print("PLAYER 1 SCORE: %d ...... PLAYER 2 SCORE: %d" % (self.players[0].get_score(), self.players[1].get_score()))
+        #commented this out because i was getting a divide by zero error
```
Comment (jfoley-yw) [inline scrabbler/simulation.py]: I think we can probably remove all of the `self.times`-related code at this point anyways.

File: scrabbler/simulation.py [hunk 6]
```diff
@@ -78,21 +82,23 @@ def generate_rack_and_bag(self):
...
-        # returned
-        before = time.time()
-        best_moves = self.game.find_valid_moves(''.join(self.racks[self.player]))
-        self.times.append(time.time() - before)
-        if len(best_moves) == 0:
-            return None
-        return best_moves[0]
+        self.players[self.player].update_rack(new_letters)
+
+    # def generate_moves(self):
```
Comment (jfoley-yw) [inline scrabbler/simulation.py]: You can delete this commented-out method.

File: scrabbler/simulation.py [hunk 7]
```diff
@@ -46,11 +48,11 @@ def simulate(self):
...
         self.generate_new_rack()
 
-        # End game once either player has no letters left
-        if not self.racks[self.player]:
+        # End of the game once either player has no letters left
+        if self.players[self.player].is_rack_empty():
             return False
 
-        best_move = self.find_best_move()
+        best_move = self.players[self.player].choose_move(self.endgame, self.game)
```
Comment (jfoley-yw) [inline scrabbler/simulation.py:61]: We probably want to pass in a deep copy of the game object here.

**Comment by sabrasisler** [comment]
Needed to update branch

**Pull request closed without merging**